Run unit tests in parallel by default.

jonathanknowles opened the pull request:
This PR adjusts the following unit test suites to run tests in parallel by default:
- `cardano-coin-selection:test:unit`
- `cardano-wallet:test:unit`
- `cardano-wallet-primitive:test:unit`
- `strict-non-empty-containers:test:unit`

It uses the `SpecHook` feature of `hspec`, as described here: https://hspec.github.io/parallel-spec-execution.html 

This makes it unnecessary to manually use the  `parallel` combinator for every individual `Spec`.

For the `cardano-wallet:test:unit` test suite in particular, this leads to a large speedup when run on a machine with at least 4 cores:

### Original timings
```
Finished in 705.3929 seconds, used 1913.9534 seconds of CPU time
real    11m49.908s
user    25m45.315s
sys     6m17.046s
```

### After using `SpecHook` to run tests in parallel by default
```
Finished in 500.0908 seconds, used 1755.7065 seconds of CPU time
real    8m25.017s
user    25m7.003s
sys     4m17.505s
```

### After removing all manual calls to `parallel` (no significant slowdown or speedup)
```
Finished in 506.2987 seconds, used 1779.8373 seconds of CPU time
real    8m31.305s
user    25m23.259s
sys     4m25.264s
```

jonathanknowles committed:
Run unit tests in parallel by default for `cardano-wallet`.

jonathanknowles committed:
Remove calls to `parallel` from unit tests for `cardano-wallet`.

jonathanknowles committed:
Run unit tests in parallel by default for `cardano-coin-selection`.

jonathanknowles committed:
Remove calls to `parallel` from unit tests for `cardano-coin-selection`.

jonathanknowles committed:
Run unit tests in parallel by default for `cardano-wallet-primitive`.

jonathanknowles committed:
Remove calls to `parallel` from unit tests for `cardano-wallet-primitive`.

jonathanknowles committed:
Run unit tests in parallel by default for `strict-non-empty-containers`.

jonathanknowles committed:
Remove calls to `parallel` from unit tests for `strict-non-empty-containers`.

paolino reviewed (approved):
Thanks, let's go (in parallel)

jonathanknowles committed:
Make `Cardano.Wallet.Shelley.NetworkSpec.observerSpec` run sequentially.

Unisay commented on lib/primitive/test/spec/Cardano/Wallet/Primitive/CollateralSpec.hs:
```diff
@@ -669,7 +669,7 @@ instance Arbitrary TxOut where
 spec :: Spec
 spec = do
-    parallel $ do
+    do
```
Mb one `do` suffice? :)

paolino replied:
parallel do

jonathanknowles replied:
> Mb one `do` suffice? :)

Nice find! 😆

jonathanknowles replied:
Fixed in 3287b9926c9e4ad517a8e47f09d20e60625455e5.

Unisay commented on lib/coin-selection/test/spec/SpecHook.hs:
```diff
@@ -0,0 +1,6 @@
+import Test.Hspec
+
+hook :: Spec -> Spec
+hook = parallel
```
👍🏼

Unisay commented on lib/wallet/test/unit/SpecHook.hs:
```diff
@@ -0,0 +1,6 @@
+module SpecHook where
```
I've noticed there is already `lib/test-utils/test/Test/Hspec/ExtraSpec.hs` for a similar purposes, do you think it makes sense to move this utility there?

jonathanknowles replied:
I suspect that module is to help with integration tests. So perhaps we could leave it untouched for the moment.

jonathanknowles committed:
One `do` is really enough in `Primitive.CollateralSpec.spec`.

This commit:

 - removes an extraneous `do` that wasn't "doing" very much.
 - removes one level of indentation that's no longer required.

In response to review feeback:

https://github.com/input-output-hk/cardano-wallet/pull/3856#discussion_r1166612817
